### PR TITLE
validation: sync chainstate to disk after syncing to tip

### LINKED ISSUE #11600: "Rolling forward" at startup can take a long time, and is not interruptible (after unclean shutdown)
bitcond is spending hours "Rolling forward" upon startup, seemingly reprocessing blocks that were previously processed. There have been reports of this happening since July on stackexchange and reddit, so I thought it time an issue was raised, given it's still happening in V0.15.1.

### PR DESCRIPTION
**Pull request by andrewtoth**
When finishing syncing the chainstate to tip, the chainstate is not persisted to disk until 24 hours after startup. This can cause [an issue where the unpersisted chainstate must be resynced if bitcoind is not cleanly shut down](https://github.com/bitcoin/bitcoin/issues/11600#issue-271011834). If using a large enough `dbcache`, it's possible the entire chainstate from genesis would have to be resynced.

This fixes the issue by persisting the chainstate to disk right after syncing to tip, but not clearing the utxo cache (using the `Sync` method introduced in https://github.com/bitcoin/bitcoin/pull/17487). This happens by scheduling a call to the new function `SyncCoinsTipAfterChainSync` every 30 seconds. This function checks that the node is out of IBD, and then checks if no new block has been added since the last call. Finally, it checks that there are no blocks currently being downloaded by peers. If all these conditions are met, then the chainstate is persisted and the function is no longer scheduled.

Mitigates #11600.

### CODE REVIEW

File: src/validation.cpp
Comment (practicalswift) [inline src/validation.cpp]: Could be written as `std::thread(FlushStateToDisk).detach();`?

Comment (practicalswift) [reply src/validation.cpp]: @andrewtoth Ah, yes! Thanks!

```
$ git grep ' FlushStateToDisk.*{'
src/validation.cpp:bool static FlushStateToDisk(const CChainParams& chainparams, CValidationState &state, FlushStateMode mode, int nManualPruneHeight) {
src/validation.cpp:void FlushStateToDisk() {
```

**Comment by laanwj** [comment]
Concept ACK, but I think `IsInitialBlockDownload` is the wrong place to implement this, as it's a query function, having it suddenly spawn a thread that flushes is unexpected.

Would be better to implement it closer to the validation logic and database update logic itself.

**Comment by andrewtoth** [comment]
@laanwj Good point. I refactored to move this behaviour to `ActivateBestChain` in an area where periodic flushes are already expected.

File: src/validation.cpp
Comment (practicalswift) [inline src/validation.cpp]: Wrong indentation :-)

Comment (andrewtoth) [reply src/validation.cpp]: Fixed.

File: src/validation.cpp
Comment (ken2812221) [inline src/validation.cpp]: I think that this variable should be the member of CChainState to avoid share same variable if we have two CChainState object in the program.

**Comment by sdaftuar** [comment]
I'm not really a fan of this change -- the problem described in #11600 is from an unclean shutdown (ie system crash), where our recovery code could take a long time (but typically would be much faster than doing a -reindex to recover, which is how our code used to work).

This change doesn't really solve that problem, it just changes the window in which an unclean shutdown could occur (reducing it at most by 24 hours).  But extra flushes, particularly during initial sync, aren't obviously a good idea, since they harm performance.  (Note that we leave IBD before we've synced all the way to the tip, I think once we're within a day or two?)

Because we flush every day anyway, it's hard for me to say that this is really that much worse, performance-wise (after all we don't currently support a node configuration where the utxo is kept entirely cached).  But I'm not sure this solves anything either, and a change like this would have to be reverted if, for instance, we wanted to make the cache actually more useful on startup (something I've thought we should do for a while).  So I think I'm a -0 on this change.

**Comment by andrewtoth** [comment]
@sdaftuar This change also greatly improves the common workflow of spinning up a high performance instance to sync, then immediately shutting it down and using a cheaper instance. Currently, you have to enter the instance and do a clean shutdown before terminating the instance. Similarly, when syncing to an external drive, you can now just unplug the drive or turn off the machine once it is synced without having to enter the machine and perform a clean shutdown.

I would argue that moving the window to 0 hours directly after initial sync is an objective improvement. There is a lot of data that will be lost directly after, so why risk another 24 hours? After that, the most they will lose is 24 hours worth of rolling back, instead of 10 years. Also, this change does *not* do any extra flushes during initial sync, only after.

I can't speak to your last point about changing the way we use the cache, since I don't know what your ideas are, but this change should be very easy to revert if future improvements would require it.

File: src/validation.cpp
Comment (andrewtoth) [reply src/validation.cpp]: Done.

**Comment by sdaftuar** [comment]
> Currently, you have to enter it and do a clean shutdown instead of just terminating.

@andrewtoth We already support this (better, I think) with the `-stopatheight` argument, no?

I don't really view data that is in memory as "at risk"; I view it as a massive performance optimization that will allow a node to process new blocks at the fastest possible speed while the data hasn't yet been flushed.  I also don't feel very strongly about this for the reasons I gave above, so if others want this behavior then so be it.

**Comment by sipa** [comment]
@sdaftuar Maybe this is a bit of a different discussion, but there is another option; namely supporting flushing the dirty state to disk, but without wiping it from the cache. Based on our earlier benchmarking, we wouldn't want to do this purely for maximizing IBD performance, but it could be done at specific times to minimize losses in case of crashes (the once per day flush for example, and also this IBD-is-finshed one).

**Comment by sdaftuar** [comment]
@sipa Agreed, I think that would make a lot more sense as a first pass optimization for the periodic flushes and would also work better for this purpose as well.

**Comment by sipa** [comment]
@andrewtoth The problem is that right now, causing a flush when exiting IBD will (temporarily) kill your performance right before finishing the sync (because it leaves you with an empty cache). If instead it was a non-clearing flush, there would be no such downside.

**Comment by sdaftuar** [comment]
My experiment in #15265 has changed my view on this a bit -- now I think that we might as well make a change like this for now, but should change the approach slightly to do something like @gmaxwell's proposal so that we don't trigger the flush before we are done syncing:

> Might it be better to trigger the flush the first time it goes 30 seconds without connecting a block and there are no queued transfers, from the scheduler thread?

**Comment by andrewtoth** [comment]
@sdaftuar @gmaxwell I've updated this to check every 30 seconds on the scheduler thread if there has been an update to the active chain height. This only actually checks after `IsInitialBlockDownload` is false, which happens if latest block is within a day of the current time. I'm not sure how to check if there are queued transfers. If this is not sufficient, some guidance on how to do that would be appreciated.

File: src/init.cpp
Comment (practicalswift) [inline src/init.cpp]: Lifetime issue with `did_initial_flush` and `last_chain_height` when `AppInitMain` goes out of scope?

Comment (andrewtoth) [reply src/init.cpp]: Moved them inside the lambda as static variables.

File: src/init.cpp
Comment (luke-jr) [inline src/init.cpp]: Can't we just unschedule the task?

Comment (andrewtoth) [reply src/init.cpp]: Unless I'm missing something, I don't think it's possible to unschedule without making a lot more changes https://github.com/bitcoin/bitcoin/blob/master/src/scheduler.h#L58

Comment (luke-jr) [reply src/init.cpp]: Use `scheduleFromNow` conditionally?

Comment (andrewtoth) [reply src/init.cpp]: Updated. How is that?

File: src/init.cpp
Comment (luke-jr) [inline src/init.cpp]: Nit: extra space on the end

File: src/init.cpp
Comment (luke-jr) [inline src/init.cpp]: I don't understand the purpose of `*` before `FlushAfterSync`.

**Review by luke-jr — APPROVED** [review]

File: src/init.cpp
Comment (andrewtoth) [reply src/init.cpp]: Removed.

File: src/init.cpp
Comment (andrewtoth) [reply src/init.cpp]: Removed them.

**Review by luke-jr — APPROVED** [review]
utACK

File: src/init.cpp
Comment (MarcoFalke) [inline src/init.cpp]: What is the purpose of scheduling a task whose only effect is to schedule itself after some time?
```suggestion
```

File: src/init.cpp
Comment (MarcoFalke) [inline src/init.cpp]: doc-nit: This comment should be moved down after the `if` or as docstring on top of the method, imo. The condition in the if below is "initial sync is *not* finished".

Comment (andrewtoth) [reply src/init.cpp]: Updated comments to align with coding style.

**Comment by MarcoFalke** [comment]
utACK 5d9aa4c6432bd41e8d65717976293e006a6c732a

**Comment by sdaftuar** [comment]
> I'm not sure how to check if there are queued transfers. If this is not sufficient, some guidance on how to do that would be appreciated.

There's a variable called `nPeersWithValidatedDownloads` in `net_processing.cpp` which indicates how many peers we are downloading blocks from.  So to implement @gmaxwell's suggestion I think you would just need to expose that variable and then check to see if it equals 0.

**Comment by andrewtoth** [comment]
@sdaftuar Thanks. Updated.

File: src/net_processing.cpp
Comment (MarcoFalke) [inline src/net_processing.cpp]: Annotations in definitions are ignored, they need to be in the declaration

```suggestion
int GetNumberOfPeersWithValidatedDownloads()
```

Comment (andrewtoth) [reply src/net_processing.cpp]: Fixed.

**Comment by TheBlueMatt** [comment]
-0.5. This is really not the right way to go here, I think. Namely, people with high dbcache may also want to use it after IBD completes (ie cause they want to connect blocks quickly), and this breaks that by wiping the cache completely. Coupling it with a change to not drop non-dirty entries while flushing is the easy fix to make it not have a regression, though a more complete cleanup of background flushing would also be nice.

**Comment by jamesob** [comment]
> this breaks that by wiping the cache completely

On a sort of related note, I think it's kind of crazy that we couple writing back to disk with emptying the in-memory cache. Feels like those should be separate considerations, especially in a case like this.

**Comment by laanwj** [comment]
This PR seems to be too controversial to merge in the current state, and discussion has become inactive. Closing as up for grabs. Let me know if you start work on this again and I'll re-open it.

**Comment by MarcoFalke** [comment]
Could rebase on top of something like: #17487, and then open a new pull request and link to this discussion which is mostly about "NACK, because this will erase the cache".

**Comment by andrewtoth** [comment]
@MarcoFalke Thanks for bringing that to my attention. What is the benefit of making a new PR instead of reopening this one? Just for visibility?

**Comment by MarcoFalke** [comment]
Up to you. Let me know if you want this reopened.

**Comment by luke-jr** [comment]
The benefit is that the no-longer-applicable criticism is no longer right there at first glance.

**Comment by MarcoFalke** [comment]
Given that it was found this does not negatively affect performance, see https://github.com/bitcoin/bitcoin/pull/15265#issuecomment-458657451, I think it should be reconsidered in its current form.

**Comment by andrewtoth** [comment]
@MarcoFalke I don't think the takeaway from that comment is that it doesn't affect performance. That was benchmarking IBD for pruned nodes with not emptying the cache on flush. This PR is flushing the dbcache after IBD. I'm sure for a non-pruned node with a high dbcache this will negatively affect performance directly after the flush, as @TheBlueMatt 's comment points out.

**Comment by andrewtoth** [comment]
If #17487 gets merged I will rebase this, perhaps without checking for active peers since the cache would no longer be affected and it will make the diff smaller.

**Comment by Sjors** [comment]
Concept ACK on doing a `Flush(erase=false)` at the end of IBD once support for that is merged. It probably still makes sense to wait a little bit (e.g. checking for active peers), otherwise the user might be bothered with this flush right before sync really finishes.

**Comment by Sjors** [comment]
> If #17487 gets merged I will rebase this

It has been merged.

File: src/net_processing.cpp
Comment (luke-jr) [inline src/net_processing.cpp:526]: Why was the order here swapped? Seems like override should be the very last keyword IMO

File: src/init.cpp
Comment (luke-jr) [inline src/init.cpp:140]: nit: There used to be a blank line before this. IMO it was nicer.

File: src/init.cpp
Comment (luke-jr) [inline src/init.cpp:1060]: Do we want to monitor only the active chainstate? Or should we be paying attention to the background one too?

File: src/init.cpp
Comment (luke-jr) [inline src/init.cpp:1069]: `-1` is always going to be `!= current_height` anyway...?

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/init.cpp
Comment (andrewtoth) [reply src/init.cpp:1060]: That's a good question.
Please correct me if my assumptions are incorrect, but I believe the other chainstate would be the background syncing while our active is the assumeutxo chainstate which we want to sync quickly and then have active ASAP. In that case, there are only two chainstates, active and background. The background chainstate can take a long time to sync, possibly >24h, which would mean the periodic flush would take effect. So syncing that here would not really be beneficial. For the active chainstate it would be much less than 24h and therefore we would want sync it to disk to protect against data loss.

**Comment by andrewtoth** [comment]
@luke-jr thank you for the speedy review! Will address your comments, but will wait for others to review as well.
@Sjors @MarcoFalke rebased, apologies for the delay.

File: src/net_processing.cpp
Comment (andrewtoth) [reply src/net_processing.cpp:526]: Hmm it seems like this is the ordering for all methods in this class definition.

**Comment by andrewtoth** [comment]
@Sjors @maflcko @luke-jr I've rebased, added some logging as well as a functional test.

**Review by Cinnamonrou — APPROVED** [review]

**Review by Cinnamonrou — APPROVED** [review]

**Review by Cinnamonrou — APPROVED** [review]

File: src/init.cpp
Comment (mzumsande) [inline src/init.cpp:2025]: Should the original scheduling be conditional on `IsInitialBlockDownload() == true`? Seems like the goal of this PR is to trigger flush if we transition out of IBD, but not so much after restarting an almost or completely synced node. Currently, it will sync 1 minute after restart (with the second call to `SyncCoinsTipAfterChainSync`, because for the first one after 30s `last_chain_height` is still -1).

**Review by mzumsande — COMMENTED** [review]
Concept ACK

While this one-time sync after IBD should help in some situations, I'm not sure that it completely resolves https://github.com/bitcoin/bitcoin/issues/11600 (I encountered this PR while looking into possible improvements to `ReplayBlocks()`)
After all, there are several other situations in which a crash / unclean shutdown could lead to extensive replays (e.g. during IBD) that this PR doesn't address.

File: src/init.cpp
Comment (chrisguida) [inline src/init.cpp:1118]: Can we clarify this message a bit to emphasize that this rescheduling does not preclude periodic flushes to disk?

It's slightly confusing to see this message as it appears that disk flushes are being deferred indefinitely.

See: https://github.com/bitcoinknots/bitcoin/issues/70#issuecomment-2136078696

Comment (andrewtoth) [reply src/init.cpp]: Updated the log message to be `"Node is still in IBD, rescheduling post-IBD chainstate disk sync..."`. Does that clarify it?

File: src/init.cpp
Comment (andrewtoth) [reply src/init.cpp]: Yes, good idea. Done.

**Comment by andrewtoth** [comment]
@mzumsande @chrisguida thank you for your reviews and suggestions. I've addressed them and rebased.

File: src/init.cpp
Comment (furszy) [inline src/init.cpp:1137]: Isn't this going to always reschedule the task on the first run?

Also, the active height refers to the latest connected block. It doesn't tell us we are up-to-date with the network. To know if we are sync, should use the best known header or call to the `ApproximateBestBlockDepth()` function.

And thinking more about this; what about adjusting the check interval based on the distance between the active chain height and the best header height?

I know this could vary a lot but.. something simple like: "if the node is more than 400k blocks away, wait 5 or 10 minutes, if it is 100k blocks away wait 3 or 5 minutes, and if it less than that, wait 1 minute" would save a good number of unneeded checks in slow machines.

File: src/init.cpp
Comment (andrewtoth) [reply src/init.cpp:1127]: Point taken for moving the explicit lock after this check, since the lock is taken in `IsInitialBlockDownload()`.

However, this check only runs once every 30 seconds. I don't see how it could possibly affect responsiveness of the software.  It is a very fast check I would assume on the order of microseconds every 30 seconds.

File: src/init.cpp
Comment (andrewtoth) [reply src/init.cpp:1137]: > Isn't this going to always reschedule the task on the first run?

Yes, but do you think this is a problem? It just makes sure the node has not connected any nodes for at least 30 seconds.

> Also, the active height refers to the latest connected block. It doesn't tell us we are up-to-date with the network; To know if we are sync, should use the best known header or call to the ApproximateBestBlockDepth() function.

Doesn't the fact that `IsInitialBlockDownload()` returns false make this point moot? It checks that our latest block is at most 24 hours old.

And let's say this call is triggered before we are completely up-to-date with the network. All that happens is the chainstate is synced to disk, but the utxo cache is not cleared. So at most 24 hours of blocks (~144 blocks) will be downloaded and processed (quickly still with the cache), but not persisted to disk until the next periodic flush (24 hours). I think this patch still achieves its goal and there is no downside now with `Sync`.

> would save a good number of unneeded checks in slow machines.

I think this is premature optimization. I don't think this check will be noticeable by system or user.

Comment (furszy) [reply src/init.cpp:1137]: Ok. np.

Still.. I know it is an overkill if we only introduce what I'm going to suggest for this PR but.. thought on adding a signal for the ibd completion state https://github.com/furszy/bitcoin-core/commit/85a050a8690e5431848658604e913c58ae45a4aa. It might be useful if we ever added other scenarios apart from this one.

**Review by furszy — COMMENTED** [review]
Code ACK 8887d28a014420668801d7e4c5d1bf45c5e93684

**Comment by andrewtoth** [comment]
Closing in favor of https://github.com/bitcoin/bitcoin/pull/30611.

**Pull request closed without merging**
